BUG: fix segfault on side-effect in __bool__ function in array.nonzero()

garyg1 opened the pull request:
Numpy should throw an exception here, instead of segfaulting. See #13631 

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

garyg1 committed:
BUG: fix segfault on side-effect in __bool__ function in array.nonzero()

Numpy should throw an exception here, instead of segfaulting.

seberg commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2289,8 +2289,15 @@ PyArray_Nonzero(PyArrayObject *self)
             for (j = 0; j < count; ++j) {
                 if (nonzero(data, self)) {
+                    if (++added_count > nonzero_count) {
+                        PyErr_SetString(PyExc_RuntimeError,
```
I do not believe you can set the errorstate without the GIL? You should probably just break and check if something odd happened after the fact. If we error at too many items read, I think we may as well error at too few items read.

eric-wieser replied:
> I think we may as well error at too few items read.

I was going to suggest it, and was surprised to see it was already fixed, so thanks for noticing!

seberg commented on numpy/core/tests/test_item_selection.py:
```diff
@@ -79,9 +79,29 @@ def test_empty_partition(self):
+        class FalseThenTrue:
+            _val = False
+            def __bool__(self):
+                try: return self._val
```
Please put the new line here, I do not think we use this style anywhere else

seberg commented on numpy/core/tests/test_item_selection.py:
```diff
@@ -79,9 +79,29 @@ def test_empty_partition(self):
+
+        try:
+            np.array([True, FalseThenTrue()]).nonzero()
+            assert_(False)
```
Probably might as well use `assert` nowadays, not sure I prefer `pass` instead of `assert True`.

eric-wieser replied:
This should be using
```
with pytest.assert_raises(RuntimeError):
    np.array([True, FalseThenTrue()]).nonzero()
```

seberg reviewed (commented):
I guess we may sometimes miss if the output has is empty (no True values), that seems fine, it just uses a single pass then, which really is also OK behaviour.

garyg1 committed:
MAINT: fix for review comments in array.nonzero()

Now we check for both over-counting and under-counting. Fixed warnings causing failing build.

eric-wieser commented on numpy/core/tests/test_item_selection.py:
```diff
@@ -79,9 +79,63 @@ def test_empty_partition(self):
+        assert_array_equal(a, b)
+
+    def test_nonzero_safety(self):
+        # from Github isue #13631
```
`gh-13631` is how we typically spell this

eric-wieser commented on numpy/core/tests/test_item_selection.py:
```diff
@@ -79,9 +79,63 @@ def test_empty_partition(self):
+        b = np.array([0, 1, 2, 3, 4, 5])
+        assert_array_equal(a, b)
+
+    def test_nonzero_safety(self):
```
Though I can see why you put it here, this test should go in `core/tests/test_numeric.py:TestNonzero` so it's with the others

garyg1 committed:
MAINT: fix for review comments

Moved np.nonzero test to core/tests/test_numeric.py. Refactor to use assert_raises().

eric-wieser commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
+
+        a = np.array([[False], [TrueThenFalse()]])
+        assert_raises(RuntimeError, np.nonzero, a)
 
```
nit: PEP8 wants an extra blank line here

eric-wieser commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
+                finally:
+                    self._val = False
+
+        a = np.array([True, FalseThenTrue()])
```
nit: could do with a comment saying "grows on the second iteration"

eric-wieser commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
+        a = np.array([[True], [FalseThenTrue()]])
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([False, TrueThenFalse()])
```
nit: could do with a comment saying "shrinks on the second iteration"

eric-wieser reviewed (approved):
Just nits left I'd be happy to address myself via the github UI, otherwise looks good. Thanks for fixing this!

garyg1 commented:
Thanks @eric-wieser  and @seberg for the helpful feedback.

garyg1 commented:
@eric-wieser and sure, please do fix the nits if you want.

eric-wieser committed:
Update test_numeric.py

seberg commented:
Thanks @garyg1 and @eric-wieser. Btw. I like the `gh-...` spelling, it comes from 8 years ago when we were moving, but then why not be specific that its github.

I suppose in principle this dance is not necessary for build-in dtypes which do not include references. But I doubt that it is noticable speed wise, so no point.
